sensor/common/compliance/nodescan_handler_test.go

```diff
@@ -0,0 +1,161 @@
+					AddedBy:   "FakeLayer",
+				},
+			},
+			LanguageComponents: nil,
+		},
+		Notes: nil,
+	}
+	return msg
+}
+
+type nodeScansProducerThatStops struct {
+	numMsgs   int
+	nodeScans chan *storage.NodeScanV2
+}
+
+type stoppable interface {
+	Stop(error)
+}
+
+func newNodeScansProducerThatStops(num int) *nodeScansProducerThatStops {
+	return &nodeScansProducerThatStops{
+		numMsgs:   num,
+		nodeScans: make(chan *storage.NodeScanV2),
+	}
+}
+
+func (nsp *nodeScansProducerThatStops) generator() <-chan *storage.NodeScanV2 {
+	return nsp.nodeScans
+}
+
+func (nsp *nodeScansProducerThatStops) start(s stoppable) {
+	go func(numMsgs int) {
+		defer close(nsp.nodeScans)
+		for i := 0; i < numMsgs; i++ {
+			nsp.nodeScans <- fakeNodeScanV2(fmt.Sprintf("Bobby_%d", i))
+			if i == 0 {
+				s.Stop(nil)
+			}
+		}
+	}(nsp.numMsgs)
+}
+
+type NodeScanHandlerTestSuite struct {
+	suite.Suite
+	mu                   *sync.Mutex
+	numReceivedAtCentral int
+	cancel               context.CancelFunc
+	ctx                  context.Context
+	toCentral            chan *central.MsgFromSensor
+}
+
+func (s *NodeScanHandlerTestSuite) SetupTest() {
+	s.mu = &sync.Mutex{}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.numReceivedAtCentral = 0
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	s.toCentral = make(chan *central.MsgFromSensor)
+}
+
+func (s *NodeScanHandlerTestSuite) TearDownTest() {
+
+}
+
+func (s *NodeScanHandlerTestSuite) startForwardingToCentral(c <-chan *central.MsgFromSensor) {
+	go func() {
+		defer close(s.toCentral)
+		for {
+			select {
+			case <-s.ctx.Done():
+				return
+			case msg, ok := <-c:
+				if !ok {
+					return
+				}
+				s.toCentral <- msg
+			}
+		}
+	}()
+}
+
+// startFakeCentral consumes all messages from s.toCentral and counts them
+func (s *NodeScanHandlerTestSuite) startFakeCentral() {
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-s.toCentral:
+				s.mu.Lock()
+				s.numReceivedAtCentral++
+				s.mu.Unlock()
+			}
+		}
+	}(s.ctx)
+}
+
+// TestStopHandler goal is to stop handler while there are still some messages to process
+// in the channel passed into NewNodeScanHandler.
+// We expect that premature stop of the handler produces no race condition or goroutine leak.
+// Exec with: go test -race -count=1 -v -run ^TestNodeScanHandler$ ./sensor/common/compliance
+func (s *NodeScanHandlerTestSuite) TestStopHandler() {
+	defer s.cancel()
+	numScans := 1000
+	// nodeScansProducerThatStops stops the handler after producing the first message.
+	p := newNodeScansProducerThatStops(numScans)
+	h := NewNodeScanHandler(p.generator())
+	p.start(h)
+	s.startFakeCentral()
+	s.startForwardingToCentral(h.ResponsesC())
+
+	err := h.Start()
+	assert.NoError(s.T(), err)
+	<-h.Stopped().Done()
+	assert.True(s.T(), h.Stopped().IsDone())
+
+	// give central a chance to receive something before we lock s.mu
+	<-time.After(50 * time.Millisecond)
```

Comment:
This seems brittle. What if after a bad refactoring (in the future) you give it more time and it receives all 1000 messages? What if 50 milliseconds are not enough to receive even 1 message on some systems?
I have no concrete ideas for improvement but it feels the test tries to do too much.

Reply:
Correct. I fixed that in a bit less flaky (but not ideal) way.